operator/pkg/ingress/secret.go

```
@@ -0,0 +1,363 @@
+		err = sm.handleSecretDeletedEvent(ev)
+	case ingressAddedEvent:
+		err = sm.handleIngressAddedEvent(ev)
+	case ingressUpdatedEvent:
+		err = sm.handleIngressUpdatedEvent(ev)
+	case ingressDeletedEvent:
+		err = sm.handleIngressDeletedEvent(ev)
+	default:
+		err = fmt.Errorf("received an unknown event: %t", ev)
+	}
+	return err
+}
+
+func (sm *syncSecretManager) handleSecretAddedEvent(ev secretAddedEvent) error {
+	if ev.secret.GetNamespace() == sm.namespace {
+		return nil
+	}
+	return sm.syncSecret(ev.secret)
+}
+
+func (sm *syncSecretManager) handleSecretUpdatedEvent(ev secretUpdatedEvent) error {
+	if ev.oldSecret.GetNamespace() == sm.namespace || ev.newSecret.GetNamespace() == sm.namespace {
+		return nil
+	}
+	return sm.syncSecret(ev.newSecret)
+}
+
+func (sm *syncSecretManager) handleSecretDeletedEvent(ev secretDeletedEvent) error {
+	if ev.secret.GetNamespace() == sm.namespace {
+		return nil
+	}
+	return sm.deleteSecret(ev.secret)
+}
+
+func (sm *syncSecretManager) handleIngressAddedEvent(ev ingressAddedEvent) error {
+	return sm.handleIngressUpsertedEvent(ev.ingress)
+}
+
+func (sm *syncSecretManager) handleIngressUpdatedEvent(ev ingressUpdatedEvent) error {
+	return sm.handleIngressUpsertedEvent(ev.newIngress)
+}
+
+// handleIngressDeletedEvent is doing nothing right now as the underlying secrets could be
+// referenced in other Ingress resources, which is common use case for wildcard secrets (e.g
+// *.foo.com)
+func (sm *syncSecretManager) handleIngressDeletedEvent(ev ingressDeletedEvent) error {
+	return nil
+}
+
+func (sm *syncSecretManager) handleIngressUpsertedEvent(ingress *slim_networkingv1.Ingress) error {
+	if !tlsEnabled(ingress) {
+		return nil
+	}
+
+	for _, tls := range ingress.Spec.TLS {
+		// check if the secret is available
+		key := getSecretKey(ingress.GetNamespace(), tls.SecretName)
+		secret, exists, err := sm.getByKey(key)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("secert does not exist: %s", key)
```

`secert` -> `secret`. Also I don't think we should throw an error here. The secret can be added afterwards.

---

corrected the typo.

If we didn't throw error, the control loop will just think that such upsert event is successfully handled (i.e. no retry is carried on if secret is added in split second).

---

also, I was thinking about having some kind of backoff retry in our control loop in general, but can't find anything like that in cilium codebase. Having backoff retry will indeed help in some cases, and it's pattern that operator-sdk provides.